src/data_source/storage/sql/transaction.rs

```diff
@@ -99,6 +108,59 @@ impl TransactionMode for Read {
             .await?;
         Ok(())
     }
+
+    fn display() -> &'static str {
+        "read-only"
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum CloseType {
+    Commit,
+    Revert,
+    Drop,
+}
+
+#[derive(Debug)]
+struct TransactionMetricsGuard<Mode> {
+    started_at: Instant,
+    metrics: PoolMetrics,
+    close_type: CloseType,
+    _mode: PhantomData<Mode>,
+}
+
+impl<Mode: TransactionMode> TransactionMetricsGuard<Mode> {
+    fn begin(metrics: PoolMetrics) -> Self {
+        let started_at = Instant::now();
+        tracing::trace!(mode = Mode::display(), ?started_at, "begin");
+        metrics.open_transactions.update(1);
+
+        Self {
+            started_at,
+            metrics,
+            close_type: CloseType::Drop,
+            _mode: Default::default(),
+        }
+    }
+
+    fn close(&mut self, t: CloseType) {
```

Comment:
Nit: Am I understanding correctly that this function sets the close type but it's the `drop` function that does the real closing operation? Can we rename this to `close_type` or `set_close`?